File: core/lua/jitaccess/core/store.lua
-- JIT Network Access - Copyright (C) 2026 Slopapalooza
-- SPDX-License-Identifier: AGPL-3.0-or-later

-- jitaccess.core.store — GrantStore + NonceStore (core/SPEC.md §4, §5)
--
-- Simple/default backend: ngx.shared dicts (process-local, worker-shared, no
-- external dependency). Hardened/shared backend: Redis — present here as
-- explicit fail-closed stubs (TODO(hardened)) so the adapters never change when
-- it's wired later.
--
-- Grants invert the ban fail-direction, so every error path denies. See
-- SECURITY-REVIEW C1/C3/H3 and core/SPEC.md §7.
--
-- NOTE: not executed on the dev host; shdict semantics validated by the harness.

local cjson = require "cjson.safe"

local _M = { _VERSION = "0.1.0" }
local methods = {}
local mt = { __index = methods }

local GRANT_PREFIX = "jit:grant:"
local NONCE_PREFIX = "jit:nonce:"

-- Constant-time string equality for the ip+cookie hash comparison.
--
-- Delegates to jitaccess.core.crypto rather than keeping a second copy: the
-- local one accumulated with `(x == y and 0 or 1)`, which branches per byte and
-- so was not actually constant-time, unlike crypto.ct_equal's branch-free
-- bor/bxor accumulator and Go's subtle.ConstantTimeCompare. Two implementations
-- of one primitive is exactly how that drift happens.
local ct_equal = require("jitaccess.core.crypto").ct_equal

-- opts.grants  : ngx.shared dict for grants        (required, Simple)
-- opts.nonces  : ngx.shared dict for spent nonces  (required, Simple) — MUST be
--                a dedicated dict, isolated from grants and from BunkerWeb bans
--                so a nonce flood can't evict grants/bans (SECURITY-REVIEW H5).
-- opts.redis   : Hardened shared backend (optional; stubbed)
-- opts.tenant  : Hardened namespace (optional)
function _M.new(opts)
  assert(opts and opts.grants and opts.nonces, "store.new: grants and nonces dicts required")
  return setmetatable({
    grants = opts.grants,
    nonces = opts.nonces,
    -- Rate-limit counters get their OWN dict. SPEC §5 requires the nonce dict to
    -- be dedicated to spent-nonce claims: sharing it let an unauthenticated
    -- attacker mint one counter per source address and LRU-evict spent-nonce
    -- markers (re-enabling replay inside the TTL) and enrollment codes.
    -- Falls back to the nonce dict so a caller that has not declared the third
    -- dict still runs, with the documented caveat.
    rl     = opts.rl or opts.nonces,
    redis  = opts.redis,     -- HARDENED: nil in Simple mode
    tenant = opts.tenant,    -- HARDENED
  }, mt)
end

-- Key schema (core/SPEC.md §4.1). For ip+cookie binding the cookie hash is part
-- of the key: without it one (service, ip) pair held exactly ONE grant, so two
-- enrolled devices behind a single NAT egress evicted each other on every knock
-- and flapped indefinitely — precisely the deployment ip+cookie exists to serve.
-- `ip` binding keeps the two-part key, so the default profile is unchanged.
local function grant_key(self, sname_canon, ip_canon, cookie_hash)
  -- TODO(hardened): if self.tenant then prefix "jit:" .. tenant .. ":grant:" ...
  local k = GRANT_PREFIX .. sname_canon .. ":" .. ip_canon
  if cookie_hash then k = k .. ":" .. cookie_hash end
  return k
end

-- ---- GrantStore ------------------------------------------------------------

-- Create/refresh a grant. ttl seconds. Returns true or nil,err (fail closed).
function methods:put_grant(sname_canon, ip_canon, rec, ttl)
  if self.redis then
    -- TODO(hardened): sign rec (mac = HMAC/AEAD(grant_sign_key, canonical(rec)))
    -- then redis SET key json EX ttl; on any error return nil (caller denies).
    return nil, "hardened redis backend not implemented"
  end
  local json = cjson.encode(rec)
  if not json then return nil, "encode failed" end
  -- A device-bound record is stored under its own cookie-qualified key.
  local ck = (rec.binding == "ip+cookie") and rec.cookie_hash or nil
  local ok, err = self.grants:set(grant_key(self, sname_canon, ip_canon, ck), json, ttl)
  if not ok then return nil, err end
  return true
end

-- Fetch the raw grant record (or nil). The kid/expiry/cookie RE-CHECK required
-- by SPEC §4.3 is composed in :is_allowed below (it needs the registry + now).
--
-- cookie_hash, when the client presented one, selects that device's own grant;
-- otherwise the ip-bound key is used. Looking the hash up (rather than scanning)
-- means a device can only ever find its OWN record.
function methods:get_grant(sname_canon, ip_canon, cookie_hash)
  if self.redis then
    -- TODO(hardened): redis GET + verify mac before trusting; error -> nil (deny).
    return nil
  end
  local json
  if cookie_hash then
    json = self.grants:get(grant_key(self, sname_canon, ip_canon, cookie_hash))
  end
  if not json then
    json = self.grants:get(grant_key(self, sname_canon, ip_canon))
  end
  if not json then return nil end
  return cjson.decode(json)   -- nil on corrupt -> treated as no grant (deny)
end

-- SPEC §4.3 is_allowed: grant present AND kid still registered AND token not
-- expired AND (ip+cookie) cookie matches. registry is duck-typed (has :lookup,
-- :is_expired). Any failure -> nil (deny). This is the function the adapter's
-- access phase calls.
--
-- Break-glass carve-out: a grant created directly by an admin over the
-- authenticated internal API (rec.manual == true) skips the registry re-check,
-- since there may be no token/kid behind it. TTL still bounds it, and revoke
-- still removes it. This is what makes the gate testable and recoverable in M1
-- before the knock protocol (M2) issues registry-backed grants.
function methods:is_allowed(sname_canon, ip_canon, registry, now, cookie_hash_present)
  local rec = self:get_grant(sname_canon, ip_canon, cookie_hash_present)
  if not rec then return nil end
  -- The key concatenates fields with ":", and a canonical server_name can itself
  -- contain ":" (a bracketed IPv6 literal canonicalizes to 2001:db8::1), so the
  -- key alone is a weak identifier. Confirm the record actually describes this
  -- request rather than trusting where it was filed.
  if rec.service ~= sname_canon or rec.ip ~= ip_canon then return nil end
  -- Check the record's OWN expiry, not just the dict TTL. shdict:set(k,v,0) means
  -- "never expire", and the BunkerWeb admin API passes ttl straight through
  -- (0 is truthy in Lua, so `tonumber(body.ttl) or 3600` keeps it), so a grant
  -- created with ttl=0 lived forever here while Go denied it on `now >= g.Exp`.
  if type(rec.exp) ~= "number" or now >= rec.exp then return nil end
  if not rec.manual then
    local token = registry and registry:lookup(rec.kid)
    if not token then return nil end               -- kid revoked/unknown -> deny
    if registry:is_expired(token, now) then return nil end
    -- Re-check the per-service allow-list too, not just registry membership.
    -- Dropping a kid from one site's allow-list while leaving the token
    -- registered (it may still serve another site) otherwise left the
    -- de-authorized device admitted for the whole grant TTL, while deleting the
    -- token evicted it at once — two admin actions that look equivalent from
    -- the console behaving differently.
    if not registry:allowed_for_service(rec.kid, sname_canon) then return nil end
  end
  if rec.binding == "ip+cookie" then
    -- The grant is additionally bound to the browser that knocked: the client
    -- must present the opaque grant-id cookie whose hash we stored. On shared
    -- egress (office NAT, CGNAT, VPN) this stops a co-located client that
    -- merely shares the IP from inheriting access (SECURITY-REVIEW C5/H11).
    if not cookie_hash_present or not rec.cookie_hash then return nil end
    if not ct_equal(cookie_hash_present, rec.cookie_hash) then return nil end
  end
  return rec
end

-- Build a grant record. `opts` = { kid, binding, cookie_hash, manual, now }.
function _M.record(sname_canon, ip_canon, kid, ttl, opts)
  opts = opts or {}
  local now = opts.now or ngx.time()
  return {
    v = 1,
    kid = kid,
    service = sname_canon,
    ip = ip_canon,
    binding = opts.binding or "ip",
    cookie_hash = opts.cookie_hash,
    manual = opts.manual or false,
    issued = now,
    exp = now + ttl,
  }
end

-- Remove EVERY grant for (service, ip): the ip-bound one and any device-bound
-- ones. An admin revoking an address means "this address loses access", not
-- "one of the devices there does" — with ip+cookie there may be several, and
-- leaving the rest behind would be a silent partial revoke.
function methods:del_grant(sname_canon, ip_canon)
  if self.redis then return nil, "hardened redis backend not implemented" end
  self.grants:delete(grant_key(self, sname_canon, ip_canon))
  -- Match each record's OWN fields, never a key prefix. Keys join fields with
  -- ":" and a canonical IPv6 address contains ":", so a prefix scan for
  -- 2001:db8::1 also deleted the unrelated client 2001:db8::1:2 — a silent
  -- over-revoke that looked like a correct one.
  for _, k in ipairs(self.grants:get_keys(0)) do
    local json = self.grants:get(k)
    if json then
      local rec = cjson.decode(json)
      if rec and rec.service == sname_canon and rec.ip == ip_canon then
        self.grants:delete(k)
      end
    end
  end
  return true
end

-- Sweep every grant for a kid (revoke a lost device — SECURITY-REVIEW H3).
-- Simple/local: scan the grant dict. get_keys is O(n) and fine at self-host
-- scale; at large scale the Hardened backend uses a bykid index instead.
function methods:revoke_token(kid)
  if self.redis then return nil, "hardened redis backend not implemented" end
  local keys = self.grants:get_keys(0)   -- all keys (home scale)
  local n = 0
  for _, k in ipairs(keys) do
    local json = self.grants:get(k)
    if json then
      local rec = cjson.decode(json)
      if rec and rec.kid == kid then
        self.grants:delete(k)
        n = n + 1
      end
    end
  end
  return n
end

function methods:list()
  if self.redis then return {} end
  local out = {}
  for _, k in ipairs(self.grants:get_keys(0)) do
    local json = self.grants:get(k)
    if json then
      local rec = cjson.decode(json)
      if rec then out[#out + 1] = rec end
    end
  end
  return out
end

-- ---- enrollment codes (single-use, TTL) ------------------------------------
-- Admin-issued one-time codes that the client exchanges at /enroll for its
-- token secret, so the secret is never placed in a QR/URL (DESIGN §6.1 / R5).
-- Stored in the dedicated nonce dict (ephemeral, single-use) under an ec: prefix.

local ENROLL_PREFIX = "jit:ec:"

function methods:enroll_code_put(code, data, ttl)
  if self.redis then return nil, "hardened redis backend not implemented" end
  local json = cjson.encode(data)
  if not json then return nil, "encode failed" end
  local ok, err = self.nonces:set(ENROLL_PREFIX .. code, json, ttl)
  if not ok then return nil, err end
  return true
end

-- Consume a code (single-use): returns its data table, or nil.
--
-- PROTOCOL §2.1 requires the code be marked consumed ATOMICALLY. A read then
-- delete is not: two requests racing across workers both `get` before either
-- `delete`, so both receive the secret and two devices end up sharing one
-- long-term key — which breaks "one token = one device" and makes regenerating
-- the token useless as a revocation.
--
-- `add` on a shared dict IS atomic across workers, so the winner of a claim
-- marker is the single consumer. Same primitive as nonce_claim below; the marker
-- outlives the code so a late loser cannot re-consume a re-issued one.
local CLAIM_PREFIX = "jit:ecx:"

-- Read a code WITHOUT consuming it, so a caller can reject one it cannot serve
-- (unknown kid) before burning it. Mirrors EnrollStore.Peek in core/go; without
-- it the Lua engines destroyed a perfectly good enrollment code on any hostile
-- POST naming a kid this instance does not have.
function methods:enroll_code_peek(code)
  if self.redis then return nil end
  local json = self.nonces:get(ENROLL_PREFIX .. code)
  if not json then return nil end
  return cjson.decode(json)
end

function methods:enroll_code_consume(code)
  if self.redis then return nil end
  local key = ENROLL_PREFIX .. code
  local json = self.nonces:get(key)
  if not json then return nil end
  -- Exactly one caller can win this; everyone else sees "exists" and gets nil.
  local won = self.nonces:add(CLAIM_PREFIX .. code, true, 900)
  if won ~= true then return nil end
  self.nonces:delete(key)
  return cjson.decode(json)
end

-- ---- NonceStore (single-use claim) -----------------------------------------

-- Atomically claim a nonce as spent. id = base64url/hex of the nonce's rand.
-- Returns true on first use, false if already spent. On Hardened backend a
-- backend error MUST fail closed (return false) — never allow-on-error.
function methods:nonce_claim(id, ttl)
  if self.redis then
    -- TODO(hardened): redis SET key 1 NX EX ttl; reply nil (set) -> first use;
    -- reply "OK" absent / error -> already spent OR unknown -> return false.
    return false   -- fail closed until implemented
  end
  -- shdict:add returns (ok, err, forcible); ok=false with err="exists" means spent.
  local ok = self.nonces:add(NONCE_PREFIX .. id, true, ttl)
  return ok == true
end

return _M

File: core/go/store.go
// JIT Network Access - Copyright (C) 2026 Slopapalooza
// SPDX-License-Identifier: AGPL-3.0-or-later

package jitcore

import (
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"sync"
)

// GrantStore + NonceStore (SPEC §4, §5) — the only stateful modules.
//
// Simple profile: in-process, mutex-guarded TTL maps. No Redis, no database.
// Because the store is process-private (nothing outside this process can write
// it) grant values need no MAC — the grant-injection threat (SECURITY-REVIEW C3)
// requires a shared writer. HARDENED/shared-backend swaps the backend behind
// these same methods and MUST then sign values and verify on read.
//
// Grants invert the usual fail direction: a stale or forged grant *opens* a dark
// service. Every path here therefore fails closed — any doubt returns nil.

const (
	BindingIP       = "ip"
	BindingIPCookie = "ip+cookie"
)

// Grant is the stored value (SPEC §4.2).
type Grant struct {
	V          int    `json:"v"`
	Kid        string `json:"kid"`
	Service    string `json:"service"` // canonical
	IP         string `json:"ip"`      // canonical
	Binding    string `json:"binding"`
	CookieHash string `json:"cookie_hash,omitempty"` // hex sha256 of the grant-id cookie
	Issued     int64  `json:"issued"`
	Exp        int64  `json:"exp"`
	Manual     bool   `json:"manual,omitempty"` // admin break-glass: skips the registry re-check
}

// CookieHash is what we persist for ip+cookie binding — never the cookie value
// itself, so a dump of the store does not yield working grant credentials.
func CookieHash(v string) string {
	sum := sha256.Sum256([]byte(v))
	return hex.EncodeToString(sum[:])
}

// GrantKey is the canonical key schema (SPEC §4.1). Kept identical across
// engines so a later shared backend interoperates with the Lua implementation.
//
// For ip+cookie binding the cookie hash is part of the key. Without it, one
// (service, ip) pair held exactly one grant, so two enrolled devices behind a
// single NAT egress evicted each other on every knock and flapped indefinitely
// — precisely the deployment ip+cookie exists to serve. `ip` binding keeps the
// two-part key, so nothing changes for the default profile.
func GrantKey(serviceCanon, ipCanon string) string {
	return "jit:grant:" + serviceCanon + ":" + ipCanon
}

// GrantKeyCookie is the ip+cookie variant: one entry per device, not per IP.
func GrantKeyCookie(serviceCanon, ipCanon, cookieHash string) string {
	return GrantKey(serviceCanon, ipCanon) + ":" + cookieHash
}

// keyFor picks the schema a grant is stored under from its own binding.
func keyFor(g *Grant) string {
	if g.Binding == BindingIPCookie && g.CookieHash != "" {
		return GrantKeyCookie(g.Service, g.IP, g.CookieHash)
	}
	return GrantKey(g.Service, g.IP)
}

type GrantStore struct {
	mu     sync.RWMutex
	grants map[string]*Grant
	byKid  map[string]map[string]bool // kid -> set of grant keys (SPEC §4.1 index)
}

func NewGrantStore() *GrantStore {
	return &GrantStore{
		grants: map[string]*Grant{},
		byKid:  map[string]map[string]bool{},
	}
}

// Put creates or refreshes a grant.
//
// oldCookieHash, when non-empty, is the hash of a grant cookie the client
// already held for this (service, ip). Because a device-bound key embeds the
// cookie hash, a re-knock that mints a FRESH cookie lands on a different key and
// would leave the previous record behind — so a browser re-knocking every grant
// TTL accumulates one dead record per knock until the sweeper runs. Passing the
// old hash retires that record in the same operation.
func (s *GrantStore) Put(g *Grant, oldCookieHash ...string) {
	if g == nil {
		return
	}
	k := keyFor(g)
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, oh := range oldCookieHash {
		if oh == "" || oh == g.CookieHash {
			continue
		}
		ok := GrantKeyCookie(g.Service, g.IP, oh)
		if prev, exists := s.grants[ok]; exists {
			s.unindexLocked(prev.Kid, ok)
			delete(s.grants, ok)
		}
	}
	if old, ok := s.grants[k]; ok {
		s.unindexLocked(old.Kid, k)
	}
	s.grants[k] = g
	if s.byKid[g.Kid] == nil {
		s.byKid[g.Kid] = map[string]bool{}
	}
	s.byKid[g.Kid][k] = true
}

// IsAllowed is the request-path check. Per SPEC §4.3 it re-validates on EVERY
// call that the grant's kid is still registered and unexpired, so revoking a
// token or letting it expire evicts live grants immediately rather than at TTL
// (SECURITY-REVIEW H3). For ip+cookie it also verifies the presented cookie.
//
// cookie is the raw grant-id cookie value ("" when absent). Returns nil = deny.
func (s *GrantStore) IsAllowed(serviceCanon, ipCanon string, reg *Registry, now int64, cookie string) *Grant {
	s.mu.RLock()
	// A device-bound grant lives under a cookie-qualified key, so try that first
	// when the client presented a cookie; fall back to the plain (ip-bound) key.
	// Looking the cookie up rather than scanning also means a device can only
	// ever find its OWN grant.
	var g *Grant
	if cookie != "" {
		g = s.grants[GrantKeyCookie(serviceCanon, ipCanon, CookieHash(cookie))]
	}
	if g == nil {
		g = s.grants[GrantKey(serviceCanon, ipCanon)]
	}
	s.mu.RUnlock()

	if g == nil || now >= g.Exp {
		return nil
	}
	// The key is built by concatenating fields with ":", and a canonical
	// server_name can itself contain ":" (a bracketed IPv6 literal canonicalizes
	// to 2001:db8::1). That makes the key alone a weak identifier, so confirm the
	// record actually describes the request rather than trusting where it was
	// filed. Cheap, and it makes any key-construction ambiguity unexploitable.
	if g.Service != serviceCanon || g.IP != ipCanon {
		return nil
	}
	// Break-glass grants created by an admin have no kid behind them; TTL and
	// explicit revoke still bound them.
	if !g.Manual {
		t := reg.Lookup(g.Kid)
		if t == nil || reg.IsExpired(t, now) {
			return nil
		}
		// The per-service allow-list is re-checked too, not just registry
		// membership. Removing a kid from one site's allow-list while leaving the
		// token registered (because it still serves another site) is a normal
		// admin action, and without this the de-authorized device kept working
		// for the whole grant TTL — while deleting the token outright evicted
		// immediately, so the two admin actions behaved differently for no
		// visible reason.
		if !reg.AllowedForService(g.Kid, serviceCanon) {
			return nil
		}
	}
	if g.Binding == BindingIPCookie {
		if cookie == "" || g.CookieHash == "" {
			return nil
		}
		if subtle.ConstantTimeCompare([]byte(CookieHash(cookie)), []byte(g.CookieHash)) != 1 {
			return nil
		}
	}
	return g
}

// Get returns a grant without policy re-checks (admin/debug only). With
// ip+cookie there may be several at one address; this returns the ip-bound one
// if present, else any device-bound one.
func (s *GrantStore) Get(serviceCanon, ipCanon string) *Grant {
	s.mu.RLock()
	defer s.mu.RUnlock()
	if g, ok := s.grants[GrantKey(serviceCanon, ipCanon)]; ok {
		return g
	}
	// Match on the record's own fields, never on a key prefix: keys join fields
	// with ":" and a canonical IPv6 address contains ":" itself, so
	// "…:2001:db8::1:" is a prefix of the DIFFERENT client "…:2001:db8::1:2".
	for _, g := range s.grants {
		if g.Service == serviceCanon && g.IP == ipCanon {
			return g
		}
	}
	return nil
}

// Revoke removes EVERY grant for (service, ip): the ip-bound one and any
// device-bound ones. An admin revoking an address means "this address loses
// access", not "one of the devices at this address does" — with ip+cookie there
// may be several, and leaving the others behind would be a silent partial
// revoke. Returns true if anything was removed.
func (s *GrantStore) Revoke(serviceCanon, ipCanon string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	removed := false
	// Field comparison, NOT a key-prefix scan. Keys join fields with ":" and a
	// canonical IPv6 address contains ":", so revoking 2001:db8::1 by prefix also
	// deleted the unrelated client 2001:db8::1:2 — a silent over-revoke that
	// looked like a correct one.
	for k, g := range s.grants {
		if g.Service == serviceCanon && g.IP == ipCanon {
			s.unindexLocked(g.Kid, k)
			delete(s.grants, k)
			removed = true
		}
	}
	return removed
}

// RevokeToken deletes every grant carrying kid and returns the count, so the
// admin sees the blast radius of revoking a lost device (SECURITY-REVIEW H3).
func (s *GrantStore) RevokeToken(kid string) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	keys := s.byKid[kid]
	n := 0
	for k := range keys {
		if _, ok := s.grants[k]; ok {
			delete(s.grants, k)
			n++
		}
	}
	delete(s.byKid, kid)
	return n
}

// List returns the live (unexpired) grants, for the admin API.
func (s *GrantStore) List(now int64) []*Grant {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]*Grant, 0, len(s.grants))
	for _, g := range s.grants {
		if now < g.Exp {
			out = append(out, g)
		}
	}
	return out
}

// Sweep drops expired records. Lazy expiry already makes them unreachable; this
// just reclaims memory and is safe to call on a timer.
func (s *GrantStore) Sweep(now int64) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	n := 0
	for k, g := range s.grants {
		if now >= g.Exp {
			s.unindexLocked(g.Kid, k)
			delete(s.grants, k)
			n++
		}
	}
	return n
}

func (s *GrantStore) unindexLocked(kid, key string) {
	if m := s.byKid[kid]; m != nil {
		delete(m, key)
		if len(m) == 0 {
			delete(s.byKid, kid)
		}
	}
}

// ---- NonceStore ------------------------------------------------------------

// NonceStore enforces single-use only (SPEC §5); the nonce authenticates itself.
// Kept separate from grants so a burst of knocks can never evict live grants
// (SECURITY-REVIEW H5).
type NonceStore struct {
	mu    sync.Mutex
	spent map[string]int64 // id -> expiry
}

func NewNonceStore() *NonceStore { return &NonceStore{spent: map[string]int64{}} }

// Claim is the atomic check-and-burn: true on first use, false if already spent.
// Called only after the proof verifies, so a bad proof cannot consume a nonce.
func (n *NonceStore) Claim(id string, now, ttl int64) bool {
	n.mu.Lock()
	defer n.mu.Unlock()
	if exp, ok := n.spent[id]; ok && now < exp {
		return false
	}
	n.spent[id] = now + ttl
	return true
}

func (n *NonceStore) Sweep(now int64) int {
	n.mu.Lock()
	defer n.mu.Unlock()
	c := 0
	for k, exp := range n.spent {
		if now >= exp {
			delete(n.spent, k)
			c++
		}
	}
	return c
}

// ---- enrollment codes ------------------------------------------------------

// EnrollCode is a single-use, TTL-bounded code exchanged at /enroll for a device
// secret, so the secret never travels in a registration URL or QR (PROTOCOL
// §6.1 / SECURITY-REVIEW R5).
type EnrollCode struct {
	Kid     string   `json:"kid"`
	Origins []string `json:"origins"`
	Exp     int64    `json:"exp"`
}

type EnrollStore struct {
	mu    sync.Mutex
	codes map[string]*EnrollCode
}

func NewEnrollStore() *EnrollStore { return &EnrollStore{codes: map[string]*EnrollCode{}} }

func (e *EnrollStore) Put(code string, c *EnrollCode) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.codes[code] = c
}

// Peek reads a code WITHOUT consuming it, so a caller can reject an unusable
// code (unknown kid, wrong service) before burning it. Consume is
// delete-after-read, so validating afterwards let any hostile POST destroy a
// perfectly good enrollment code.
func (e *EnrollStore) Peek(code string, now int64) *EnrollCode {
	e.mu.Lock()
	defer e.mu.Unlock()
	c, ok := e.codes[code]
	if !ok || now >= c.Exp {
		return nil
	}
	return c
}

// Consume is delete-after-read: a code works exactly once.
func (e *EnrollStore) Consume(code string, now int64) *EnrollCode {
	e.mu.Lock()
	defer e.mu.Unlock()
	c, ok := e.codes[code]
	if !ok {
		return nil
	}
	delete(e.codes, code)
	if now >= c.Exp {
		return nil
	}
	return c
}

func (e *EnrollStore) Sweep(now int64) int {
	e.mu.Lock()
	defer e.mu.Unlock()
	n := 0
	for k, c := range e.codes {
		if now >= c.Exp {
			delete(e.codes, k)
			n++
		}
	}
	return n
}
